Additional features to Comment section

nif33 opened the pull request:
Add option for viewing a limited number or comments
Add button to delete all comments

nif33 committed:
Allow users to choose number of displayed comments

add form in HTML for users to select display number
servlet only gets the required number of comments
if no input or invalid input, defaults to display 5 comments

nif33 committed:
Allow users to choose number of displayed comments

add form in HTML for users to select display number
servlet only gets the required number of comments
if no input or invalid input, defaults to display 5 comments

nif33 committed:
Add delete comments feature

deletes all comments in the database
removes all comments currently displayed
asks user to verify choosing to delete

nif33 committed:
Add delete comments feature

deletes all comments in the database
removes all comments currently displayed
asks user to verify choosing to delete

nif33 committed:
format comment section to match rest of website

nif33 committed:
format comment section to match rest of website

nif33 commented on portfolio/src/main/webapp/script.js:
```diff
@@ -48,12 +48,20 @@ function newCommentBox(comment){
+  const commentContainer = document.getElementById('comment-container');
+  const parameters = location.search;
+  // Clear previous children
+  while(commentContainer.lastElementChild) {
```
I'm not sure what the best way to clear the comment container dom is. I found this resource https://stackoverflow.com/questions/3955229/remove-all-child-elements-of-a-dom-node-in-javascript which lists a bunch of options and chose looping + deleting.

juanalonsocruz replied:
Yea I tried searching this earlier. I don't mind this. I wonder why there isn't a simple method for this.

nif33 commented on portfolio/src/main/webapp/script.js:
```diff
@@ -48,12 +48,20 @@ function newCommentBox(comment){
+  }
+
+  // Repopulate comment section
+  fetch('/data'+parameters).then(response => response.json()).then((comments) => {
```
I wrote this before the discussion on using the Request class today. I know now that the Request class makes it more readable, but I think this method is pretty readable too? I guess the main difference is it does not state the name of the parameters and just passes them. Interested to know what you think.

juanalonsocruz replied:
Correct. It is hard to tell what you parameters you are trying to pass. Assuming parameters was a nice simple string, then I guess its fine. In that case, Andrew would recommend template strings ```fetch(`/data?max-comments=${maxComments}`)```

With more complicated parameters, I would recommend Request.

juanalonsocruz commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -59,8 +59,25 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
+      numCommentsToDisplay = Integer.parseInt(request.getParameter("limit"));
+      System.err.println(request.getParameter("limit"));
+    } catch (NumberFormatException e) {
+      System.err.println("Input for limit invalid. Displaying default number.");
+      numCommentsToDisplay = 5;
```
Another way to handle a bad input, is to not execute the rest of the servlet and return an HTTP error. Not required for this starter project but something useful that you will probably use a lot in the capstone.

https://tomcat.apache.org/tomcat-5.5-doc/servletapi/javax/servlet/http/HttpServletResponse.html#sendError(int,%20java.lang.String)

juanalonsocruz commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -59,8 +59,25 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
+    int numCommentsToDisplay;
+    try {
+      numCommentsToDisplay = Integer.parseInt(request.getParameter("limit"));
+      System.err.println(request.getParameter("limit"));
```
Remove this. There is no error happening here.

juanalonsocruz commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -59,8 +59,25 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
+    }
+
     ArrayList<Comment> comments = new ArrayList<>();
     for (Entity entity : results.asIterable()) {
+      // Check whether has reached comment display limit
+      if(numCommentsToDisplay == 0) {
+        break;
+      }
+      numCommentsToDisplay--;
```
We can make this more efficient by making the query return a specific number of elements. Right now you are getting all the data from the database into the servlet. Imagine your database had millions of entries. Then in this case they would all be loaded into the servlet. Well actually I'm not sure how PreparedQuery.asIterable() is implemented but its good practice to make your query for any database do minimal work needed.

To fix this, look into `FetchOptions fetchOptions = FetchOptions.Builder.withLimit(maxComments);`

juanalonsocruz commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -59,8 +59,25 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
+    // Get number of comments to display
+    int numCommentsToDisplay;
+    try {
+      numCommentsToDisplay = Integer.parseInt(request.getParameter("limit"));
```
This limit could negative so you also want to handle negative numbers. 

Andrew has mentioned in other's PRs:

> It's also a good idea in general to set a maximum number of comments.
> Right now I can crash your server by requesting more comments than your server can hold in memory (assuming there are actually that many in the database)

juanalonsocruz commented on portfolio/src/main/webapp/index.html:
```diff
@@ -57,31 +57,44 @@ <h2>My Favourite Place</h2>
+      </form>
       <div id="comment-container"></div>
 
       <form action="/data" method="POST">
```
Notice how when you set your comment limit to something that is not the default and then you submit a new comment, it goes back to the default limit. This is because your forcing the user to go to the /data page and then the servlet forces them back. You can instead use a POST fetch.

juanalonsocruz commented on portfolio/src/main/webapp/script.js:
```diff
@@ -47,17 +47,55 @@ function newCommentBox(comment){
 function loadComments() {
-  fetch('/data').then(response => response.json()).then((comments) => {
+  const commentContainer = document.getElementById('comment-container');
+  const parameters = location.search;
```
What is location.search used for here? Should this be the limit param?

nif33 replied:
location.search here is the entire query string. Should I change the variable "parameters" to "queryString" to make that more readable? Or is it bad practice to use location.search since it passes all the parameters to fetch at once? Here is where I found location.search https://www.w3schools.com/jsref/prop_loc_search.asp

juanalonsocruz replied:
It is not obvious to me what parameters are included in location.search that you care about that. I'm guessing it is the limit parameter. I would think you would get it from the selector element. Am I missing something?

nif33 replied:
ohh okay I get it now, thanks!

juanalonsocruz commented on portfolio/src/main/webapp/script.js:
```diff
@@ -47,17 +47,55 @@ function newCommentBox(comment){
   });
 }
 
+function verifyDelete() {
+  var inputWord = prompt('What\'s the magic word?');
+  if (inputWord == 'please') {
+    alert('Comments have been deleted.')
+    return true;
+  } else {
+    alert('That word is not magic.');
+    return false;
+  }
+}
```
lol

Nit: I shared with all of you one of my favorite ToTTs. It basically talks how the less indentation you have the better. In this case you don't need an else block because you return inside the if block.

nif33 committed:
change comment input form to a POST fetch

the comment submission used to redirect to the /data page and back to /index
changing because refreshing would revert the user's choice in comment showing limit

nif33 committed:
fix conflicts with remote branch

juanalonsocruz commented on portfolio/src/main/webapp/index.html:
```diff
@@ -79,16 +79,16 @@ <h2 id="comment-section">Comments</h2>
         <br/><br/>
 
-        <input type="submit"/>
+        <input type="button" value="Search" onclick="return addComment()"/>
```
Do you need `return` ?

juanalonsocruz commented on portfolio/src/main/webapp/script.js:
```diff
@@ -73,6 +73,23 @@ function loadComments() {
+    {method: 'POST'}
+  );
+  fetch(request).then(() => {
+    const newComment = {name:nameInput, text:commentInput};
+    commentContainer.appendChild(newCommentBox(newComment));
```
It is better practice to just call `loadComments()` here. If the back-end has a bug and doesn't actually store it in the database, then here it looks like it successfully worked. Just in generally, best to show canonical data.

juanalonsocruz reviewed (approved):
Approving but expecting changes from comments.

ajpaon commented on portfolio/src/main/webapp/script.js:
```diff
@@ -47,17 +47,100 @@ function newCommentBox(comment){
   });
 }
 
+function verifyDelete() {
```
Readability comment: it's not clear from the name of this function how I'm expected to use it. I need to read the function in order to figure out that it returns boolean, and I need to read it closely to figure out that it returns true when I should delete.

For comparison, here is another approach to implementing a verification function

```
function verifyDelete() {
 var inputWord = prompt('What\'s the magic word?');
  if (inputWord != 'please') {
    alert('Comments have been deleted.')
    throw new Error(`User entered ${inputWord}`);
  }
}
```

Then usage would be totally different:

```
function deleteComments() {
  verifyDelete();  // if this throws an exception, we don't execute the rest of the function

  // implementation
}
```

So instead of calling the function "verifyDelete", maybe we can name it something like "shouldDelete". Then it's clear that the function returns a boolean, and if it returns true then we "should delete" it.

nif33 committed:
improve readability

change function name from verifyDelete to shouldDelete
since the function returns a boolean
remove unnecessary return statement in HTML
change adding new comment to HTML in addComment to use the
loadComments() function
